commit-status-publisher-server/src/main/java/jetbrains/buildServer/commitPublisher/CommitStatusPublisherListener.java

```diff
@@ -232,9 +231,9 @@ private void runForEveryPublisher(@NotNull Event event, @NotNull SBuildType buil
       }
       myProblems.clearProblem(publisher);
       for (BuildRevision revision : revisions) {
-        if (!publishedRevisions.contains(revision)) {
+        if (!publishedRevisionsByPublisher.contains(pubEntry.getKey(), revision)) {
           runTask(event, build.getBuildPromotion(), description, task, publisher, revision);
-          publishedRevisions.add(revision);
+          publishedRevisionsByPublisher.add(pubEntry.getKey(), revision);
```

Comment:
It seems to me using build feature id here is a bit of an overkill. You have covered a corner case of having two significantly different publishers for the same revision, but now all commit status publishers of the same type that happen to fire on the same revision will publish (same) statuses (multiple times).

I guess a better heuristic would be to use a publisher type instead of a build feature id. It will leave one strange corner case out - when you have, say, two instances of Upsource looking at the same VCS root, but this does not come across my mind as a realistic scenario, so it is probably ok to leave that one out.

Reply:
First part of key replaced by commit status publisher ID